docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,258 @@
+
+2. **`garbage-collect`**
+   - Scans the object store for older snapshots and deletes them if their immutability period has expired and they are no longer needed, following the standard [garbage collection policy](https://github.com/gardener/etcd-backup-restore/blob/master/docs/usage/garbage_collection.md).
+
+##### etcd Controller Enhancements
+
+When a hibernation flow is initiated (by external tooling or higher-level operators), the [etcd controller](https://github.com/gardener/etcd-druid/blob/master/docs/development/controllers.md#etcd-controller) can:
+
+1. Remove etcd’s client ports (2379/2380) from the etcd Service to block application traffic.
+2. Trigger an [on-demand full snapshot](https://github.com/gardener/etcd-druid/blob/master/docs/proposals/05-etcd-operator-tasks.md#trigger-on-demand-fulldelta-snapshot) via an `EtcdOperatorTask`.
+3. Scale down the `StatefulSet` replicas to zero, provided the previous snapshot step is successful.
+4. Create the `ExtendFullSnapshotImmutabilityTask` if `etcd.spec.backup.store.immutability.retentionType` is `"Bucket"` and based on `etcd.spec.backup.fullSnapshotSchedule`.
+
+##### Operator Task Enhancements
+
+The `ExtendFullSnapshotImmutabilityTask` will create a cron job that:
+
+- Runs `etcdbrctl reupload-snapshot` to extend the immutability of the most recent snapshot.
+- Runs `etcdbrctl garbage-collect --garbage-collection-policy <policy>` to remove old snapshots.
+
+By periodically re-uploading the latest snapshot during hibernation, the authors ensure that the immutability period is extended, and the backups remain **protected throughout the hibernation period**.
+
+###### Example Task Config
+
+```go
+type ExtendFullSnapshotImmutabilityTaskConfig struct {
+  // Schedule defines a cron schedule (e.g., "0 */6 * * *").
+  Schedule *string `json:"schedule,omitempty"`
+
+  // GarbageCollectionConfig specifies the configuration for snapshot GC.
+  GarbageCollectionConfig *GarbageCollectionConfig `json:"garbageCollectionConfig,omitempty"`
+}
+
+type GarbageCollectionConfig struct {
+  // GarbageCollectionPolicy (e.g., "LimitBased" or "Exponential").
+  GarbageCollectionPolicy *string `json:"garbageCollectionPolicy,omitempty"`
+  // MaxBackupsLimitBasedGC sets the maximum number of full snapshots to keep.
+  MaxBackupsLimitBasedGC *int32 `json:"maxBackupsLimitBasedGC,omitempty"`
+  // DeltaSnapshotRetentionPeriod indicates how long to keep delta snapshots (e.g., "72h").
+  DeltaSnapshotRetentionPeriod *metav1.Duration `json:"deltaSnapshotRetentionPeriod,omitempty"`
+}
+```
+
+**Sample YAML**:
+
+```yaml
+spec:
+  config:
+    schedule: "0 0 * * *"
+    garbageCollectionConfig:
+      garbageCollectionPolicy: "LimitBased"
+      maxBackupsLimitBasedGC: 5
+      deltaSnapshotRetentionPeriod: "72h"
+```
+
+## Compatibility
+
+These changes are compatible with existing etcd clusters and current backup processes.
+
+- **Backward Compatibility:**
```

Comment:
What about reverting from immutable backups to mutable backups? Is that possible? If so, can you please elaborate on the nuances of this? If not, then can you add that as a limitation here?

Reply:
Yes, added a section for disabling the immutability [here](https://github.com/gardener/etcd-druid/pull/884/commits/6e4511a5afcb65ebe8ef391826b7ec1769411853#diff-981f3bcbeed277184597de9af59f3e40fa0c76a43f84c739a0d17fd3948d2f8aR253-R263)